neard/src/genesis_validate.rs

```diff
@@ -1,104 +1,133 @@
-use near_chain_configs::Genesis;
+use near_chain_configs::{Genesis, GenesisConfig};
 use near_crypto::key_conversion::is_valid_staking_key;
 use near_primitives::state_record::StateRecord;
 use num_rational::Rational;
 use std::collections::{HashMap, HashSet};
 
-/// Validate genesis config and records. Panics if genesis is ill-formed.
-pub fn validate_genesis(genesis: &Genesis) {
-    let validators = genesis
-        .config
-        .validators
-        .clone()
-        .into_iter()
-        .map(|account_info| {
-            assert!(
-                is_valid_staking_key(&account_info.public_key),
-                "validator staking key is not valid"
-            );
-            (account_info.account_id, account_info.amount)
-        })
-        .collect::<HashMap<_, _>>();
-    assert_eq!(
-        validators.len(),
-        genesis.config.validators.len(),
-        "Duplicate account in validators"
-    );
-    assert!(!validators.is_empty(), "no validators in genesis");
-
-    let mut total_supply = 0;
-    let mut staked_accounts = HashMap::new();
-    let mut account_ids = HashSet::new();
-    let mut access_key_account_ids = HashSet::new();
-    let mut contract_account_ids = HashSet::new();
-    for record in genesis.records.0.iter() {
+pub struct GenesisValidator<'a> {
```

Comment:
```suggestion
struct GenesisValidator<'a> {
```

In general, I find it very helpful to keep visibility as low as possible -- it makes reading the code later easier, as you can see which is public interface, and which are impl details.